File: tests/test_model_io.py
"""
Test model import/export functionality
"""

import os
import numpy as np
import pytest
from gbrs import GBRS, save_predictions, load_predictions

DATA_DIR = os.path.join("tests", "data")
os.makedirs(DATA_DIR, exist_ok=True)


def test_python_save_load():
    """Test saving and loading models in Python"""
    print("Testing Python model save/load...")

    # Create simple dataset
    np.random.seed(42)
    X = np.random.randn(100, 3)
    y = X[:, 0] + 2 * X[:, 1] - X[:, 2] + np.random.randn(100) * 0.1

    # Fit model
    model = GBRS(n_iter=10, lr=0.1, n_quantiles=5)
    model.fit(X, y)

    # Save model using method
    model.save_model(os.path.join(DATA_DIR, "test_model_python.json"))
    print("✓ Model saved to test_model_python.json")

    # Load model using class method
    loaded_model = GBRS.load_model(os.path.join(DATA_DIR, "test_model_python.json"))
    print("✓ Model loaded")

    # Verify loaded model predictions match original
    loaded_preds = loaded_model.predict(X)
    original_preds = model.predict(X)

    assert np.allclose(
        original_preds, loaded_preds
    ), "Loaded model predictions don't match original!"
    print("✓ Loaded model predictions match original!")

    # Make predictions and save them
    predictions = model.predict(X)
    save_predictions(predictions, os.path.join(DATA_DIR, "predictions_python.json"))
    print(f"✓ Predictions saved: {len(predictions)} values")

    # Load predictions
    loaded_preds_json = load_predictions(
        os.path.join(DATA_DIR, "predictions_python.json")
    )
    print(f"✓ Predictions loaded: {len(loaded_preds_json)} values")

    # Verify predictions match
    assert np.allclose(predictions, loaded_preds_json), "Predictions don't match!"
    print("✓ Predictions match!")

    print("\n✅ All Python save/load tests passed!\n")


if __name__ == "__main__":
    test_python_save_load()

File: tests/test_model_io.R
# Test model import/export functionality for R

library(gbrs)

test_r_save_load <- function() {
  cat("Testing R model save/load...\n")
  
  # Use mtcars dataset
  set.seed(42)
  
  # Fit model
  model <- gbrs(mpg ~ wt + hp, data = mtcars, n_max = 10, lr = 0.1)
  cat("✓ Model fitted\n")
  
  # Save model
  save_gbrs(model, "test_model_r.json")
  cat("✓ Model saved to test_model_r.json\n")
  
  # Load model
  loaded_model <- load_gbrs("test_model_r.json")
  cat(sprintf("✓ Model loaded: %d rules\n", nrow(loaded_model$weights)))
  
  # Make predictions and save them
  predictions <- predict(model, mtcars)
  save_predictions(predictions, "predictions_r.json")
  cat(sprintf("✓ Predictions saved: %d values\n", length(predictions)))
  
  # Load predictions
  loaded_preds <- load_predictions("predictions_r.json")
  cat(sprintf("✓ Predictions loaded: %d values\n", length(loaded_preds)))
  
  # Verify predictions match (within numerical tolerance)
  cat("Checking predictions equality...\n")
  # Strip names for comparison (loaded predictions don't have names)
  preds_no_names <- as.numeric(predictions)
  loaded_no_names <- as.numeric(loaded_preds)
  max_diff <- max(abs(preds_no_names - loaded_no_names))
  tolerance <- 1e-4  # JSON serialization precision limit
  if (max_diff < tolerance) {
    cat(sprintf("✓ Predictions match (max diff: %.2e, tolerance: %.2e)\n", max_diff, tolerance))
  } else {
    stop(sprintf("Predictions differ too much! Max diff: %.2e (tolerance: %.2e)", max_diff, tolerance))
  }
  
  cat("\n✅ All R save/load tests passed!\n\n")
}

# Run test
test_r_save_load()
